libexec/rtld-elf/rtld_c18n.h

```diff
@@ -45,113 +44,55 @@ extern const char *ld_compartment_unwind;
 /*
  * Policies
  */
+/*
+ * RTLD is the first compartment.
+ */
+#define	RTLD_COMPART_ID		0
+
 typedef uint16_t compart_id_t;
-#define	C18N_COMPARTMENT_ID_MAX	(UINT16_MAX >> 1)
+/*
+ * Define another type for the stack table index to avoid confusion.
+ */
+typedef struct { uint16_t val; } stk_table_index;
 
 compart_id_t compart_id_allocate(const char *);
 
 /*
  * Stack switching
  */
-/*
- * This macro can only be used in a function directly invoked by a trampoline.
- */
-#define	get_trusted_stk()	({					\
-		struct trusted_frame *_tf;				\
-		_Pragma("clang diagnostic push");			\
-		_Pragma("clang diagnostic ignored \"-Wframe-address\"");\
-		_tf = __builtin_frame_address(1);			\
-		_Pragma("clang diagnostic pop");			\
-		_tf;							\
-	})
-
 struct stk_table {
 	union {
-		void *(*resolver)(unsigned);
+		void (*resolver)(void);
 		SLIST_ENTRY(stk_table) next;
 	};
-	size_t capacity;
-	struct stk_table_stack {
-		void *bottom;
-		size_t size;
-	} stacks[];
+	struct {
```

Comment:
Please name structs like this, and avoid nesting, especially when there are pointers to these structs involved.